### ISSUE TITLE
Continue issue with _GLFW_WNDCLASSNAME

### ISSUE THREAD

**Issue by irov** [glfw/glfw#1395]
in this issue we add define for WNDCLASSNAME https://github.com/glfw/glfw/issues/1120

It is time to add this setting in CMakeLists.txt as option

Now it is impossible to change this define via **ExternalProject_Add** which is used in **CI**

If you need a pull request, then I can prepare it

**Comment by jjYBdx4IL** [comment]
It would be nice to have GLFW as a thin wrapper around DLL plugins/modules for cross-platform compatibility.

BTW: the problem isn't the window class name per se (unique per HINSTANCE). It's rather that GLFW gets the HINSTANCE for the main process into which it gets loaded into. Instead, it should use the module handle:

```
diff --git a/include/GLFW/glfw3native.h b/include/GLFW/glfw3native.h
index e680c1e3..9154c711 100644
--- a/include/GLFW/glfw3native.h
+++ b/include/GLFW/glfw3native.h
@@ -169,6 +169,13 @@ GLFWAPI const char* glfwGetWin32Monitor(GLFWmonitor* monitor);
  *  @ingroup native
  */
 GLFWAPI HWND glfwGetWin32Window(GLFWwindow* window);
+
+// Plugins need to set this before using any parts of GLFW to avoid clashes
+// with other plugins potentially also using GLFW (also, DLLs shouldn't
+// register window classes using the EXE's instance handle). This should
+// be more in accordance with Win32 rules than the redefinition of
+// _GLFW_WNDCLASSNAME. It's also more flexible.
+GLFWAPI void glfwSetHinstance(void* hInst);
 #endif

 #if defined(GLFW_EXPOSE_NATIVE_WGL)
diff --git a/src/init.c b/src/init.c
index a7b5af63..0d1532c1 100644
--- a/src/init.c
+++ b/src/init.c
@@ -41,6 +41,10 @@
 //
 // Any other global variable is a bug

+#if defined(_GLFW_WIN32)
+void* hInstance = NULL;
+#endif
+
 // Global state shared between compilation units of GLFW
 //
 _GLFWlibrary _glfw = { GLFW_FALSE };
diff --git a/src/internal.h b/src/internal.h
index f87c2f93..5264d59d 100644
--- a/src/internal.h
+++ b/src/internal.h
@@ -518,6 +518,10 @@ struct _GLFWmutex
     _GLFW_PLATFORM_MUTEX_STATE;
 };

+#if defined(_GLFW_WIN32)
+void* __hInstance;
+#endif
+
 // Library global data
 //
 struct _GLFWlibrary
diff --git a/src/win32_init.c b/src/win32_init.c
index 970da06e..e292bc9e 100644
--- a/src/win32_init.c
+++ b/src/win32_init.c
@@ -68,6 +68,13 @@ BOOL WINAPI DllMain(HINSTANCE instance, DWORD reason, LPVOID reserved)

 #endif // _GLFW_BUILD_DLL

+#if defined(_GLFW_WIN32)
+void glfwSetHinstance(void* hInst)
+{
+    __hInstance = hInst;
+}
+#endif
+
 // Load necessary libraries (DLLs)
 //
 static GLFWbool loadLibraries(void)
@@ -348,7 +355,7 @@ static GLFWbool createHelperWindow(void)
                         WS_CLIPSIBLINGS | WS_CLIPCHILDREN,
                         0, 0, 1, 1,
                         NULL, NULL,
-                        GetModuleHandleW(NULL),
+                        __hInstance != NULL ? __hInstance : GetModuleHandleW(NULL),
                         NULL);

     if (!_glfw.win32.helperWindowHandle)
diff --git a/src/win32_joystick.c b/src/win32_joystick.c
index 9c71d114..12068d89 100644
--- a/src/win32_joystick.c
+++ b/src/win32_joystick.c
@@ -574,7 +574,7 @@ GLFWbool _glfwPlatformInitJoysticks(void)
 {
     if (_glfw.win32.dinput8.instance)
     {
-        if (FAILED(DirectInput8Create(GetModuleHandle(NULL),
+        if (FAILED(DirectInput8Create(__hInstance != NULL ? __hInstance : GetModuleHandleW(NULL),
                                       DIRECTINPUT_VERSION,
                                       &IID_IDirectInput8W,
                                       (void**) &_glfw.win32.dinput8.api,
diff --git a/src/win32_window.c b/src/win32_window.c
index 52a9c680..3ece4503 100644
--- a/src/win32_window.c
+++ b/src/win32_window.c
@@ -1266,7 +1266,7 @@ static int createNativeWindow(_GLFWwindow* window,
                                            fullWidth, fullHeight,
                                            NULL, // No parent window
                                            NULL, // No window menu
-                                           GetModuleHandleW(NULL),
+                                            __hInstance != NULL ? __hInstance : GetModuleHandleW(NULL),
                                            (LPVOID) wndconfig);

     free(wideTitle);
@@ -1355,12 +1355,12 @@ GLFWbool _glfwRegisterWindowClassWin32(void)
     wc.cbSize        = sizeof(wc);
     wc.style         = CS_HREDRAW | CS_VREDRAW | CS_OWNDC;
     wc.lpfnWndProc   = (WNDPROC) windowProc;
-    wc.hInstance     = GetModuleHandleW(NULL);
+    wc.hInstance     = __hInstance != NULL ? __hInstance : GetModuleHandleW(NULL);
     wc.hCursor       = LoadCursorW(NULL, IDC_ARROW);
     wc.lpszClassName = _GLFW_WNDCLASSNAME;

     // Load user-provided icon if available
-    wc.hIcon = LoadImageW(GetModuleHandleW(NULL),
+    wc.hIcon = LoadImageW(__hInstance != NULL ? __hInstance : GetModuleHandleW(NULL),
                           L"GLFW_ICON", IMAGE_ICON,
                           0, 0, LR_DEFAULTSIZE | LR_SHARED);
     if (!wc.hIcon)
@@ -1385,7 +1385,7 @@ GLFWbool _glfwRegisterWindowClassWin32(void)
 //
 void _glfwUnregisterWindowClassWin32(void)
 {
-    UnregisterClassW(_GLFW_WNDCLASSNAME, GetModuleHandleW(NULL));
+    UnregisterClassW(_GLFW_WNDCLASSNAME, __hInstance != NULL ? __hInstance : GetModuleHandleW(NULL));
 }


@@ -2377,7 +2377,7 @@ VkResult _glfwPlatformCreateWindowSurface(VkInstance instance,

     memset(&sci, 0, sizeof(sci));
     sci.sType = VK_STRUCTURE_TYPE_WIN32_SURFACE_CREATE_INFO_KHR;
-    sci.hinstance = GetModuleHandle(NULL);
+    sci.hinstance = __hInstance != NULL ? __hInstance : GetModuleHandleW(NULL);
     sci.hwnd = window->win32.handle;

     err = vkCreateWin32SurfaceKHR(instance, &sci, allocator, surface);

```

You could rename this ticket to "add module encapsulation support". For that purpose, I also wonder if there are any other such "transgressions" hidden in the GLFW code.

**Comment by elmindreda** [comment]
This should be fixed now with 07a5518c3e42f4412a5d9fa8620950d8f5041e28, which works basically like the suggestion by @jjYBdx4IL above. Thanks!
